Update expected illink warnings

sbomer opened the pull request:
~~This fixes the warning tests to also check for missing analysis warnings and updates the expected warnings to match what is actually produced.~~ edit: we want to ignore missing warnings. Now the change is only updating the list of expected warnings.

@tlakollo I think we want it to fail for missing warnings too so that we catch any that "accidentally" disappear - even if it means a bit more churn as we update framework annotations.

@eerhardt would you expect that the ComponentModel ones are gone due to https://github.com/dotnet/runtime/pull/46964? I don't understand why these would be in the console output in the first place, and I don't see anything obvious in your change that would have removed them. I wonder if the baselines were taken from the netcoreapp linker output instead of the console output.

@agocke the NullableTypeInfo warnings are gone as expected :)

sbomer committed:
Check for missing trim analysis warnings

This fixes the warning tests to also check for missing analysis warnings
and updates the expected warnings to match what is actually produced.

tlakollo commented:
I wrote code to fail on missing warnings, I guess they were changing the warnings a lot and remove that piece of code since is still explained in the comment on why a list of strings is used instead of an xml. But now we have a more stable number of warnings, I guess is ok to check again for missing warnings :D

eerhardt commented:
> I think we want it to fail for missing warnings too so that we catch any that "accidentally" disappear - even if it means a bit more churn as we update framework annotations.

We explicitly removed this in https://github.com/dotnet/sdk/pull/15419#issuecomment-763873877. We don't want this test failing and blocking runtime ingestion.

agocke commented:
@eerhardt Thanks for the insight. I was mistakenly using that test as a benchmark for how many warnings we have left to remove for console app. Given how few we have left, I suspect it won't matter for much longer! 😄

sbomer committed:
Add back warnings for Windows

The ComponentModel warnings show up due to a COM scenario on windows.

The LicenseProvider getter shows up in "copyused" but not "link" because
it is only considered an unreachable body for "link".

sbomer committed:
Don't check for missing warnings

We had decided not to check to avoid blocking dependency updates.

tlakollo commented on src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs:
```diff
@@ -502,11 +509,10 @@ private void ValidateWarningsOnHelloWorldApp (PublishCommand publishCommand, Com
             //since the error experience is not good for a developer
```
Might be good to change this comment since it no longer talks about the behavior of the code

sbomer commented:
Thanks for the additional context @eerhardt - I removed the check for missing warnings. For anyone following along:
- Those CompontentModel warnings only show up on windows because this is a COM activation scenario.

I looked into the warning behavior we didn't understand @eerhardt:

- Why does the warning about` LicenseProviderAttribute.LicenseProvider.get` show up with copyused but not link?
  - It was because we consider the getter to be an unreachable body when using link.
    
    [This](https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivator.cs#L692) reflection call causes LicenseProviderAttribute to be marked. For copyused, this also tracks the type as instantiated (marks the ctor), but for link it does not. Then when we mark the [getter](https://github.com/dotnet/runtime/blob/main/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs#L289), for link we never see the LicenseProvider instantiated, so the linker removes the method body.

- Why did the ComponentModel warnings only recently show up in the tests? (https://github.com/dotnet/sdk/pull/15963/commits/47762790ffdd767746b3a97f02d3b553a3a8bee5)
  - They weren't showing up due to the issue with \<AssemblyName\>.xml (https://github.com/mono/linker/issues/1836#issuecomment-780931488), but when @mateoatr fixed that they started showing up. Phew! :)

sbomer committed:
Update comment

agocke reviewed (approved):
LGTM

eerhardt reviewed (approved):
Looks good! Thanks for the investigation here @sbomer.
